Encryption.md

```diff
@@ -75,55 +75,126 @@ written to the output stream.
 operations are implemented in software, and can be somewhat slow, becoming a performance 
 bottleneck in certain workloads. AES-CTR is a regular (not authenticated) cipher.
 It is faster than AES-GCM, since it doesn’t perform integrity verification and doesn’t 
-calculate the authentication tag. For applications running without AES acceleration and 
+calculate the authentication tag. Actually, AES-GCM is a combination of AES-CTR and an 
+authentication module called GMAC. For applications running without AES acceleration and 
 willing to compromise on content verification, AES-CTR can provide a boost in Parquet 
 write/read throughput. The second Parquet algorithm encrypts the data content (pages) 
 with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
 the bulk of the data faster, while still verifying the metadata integrity and making sure 
-the file had not been replaced with a wrong version. However, tampering with the page data 
+the file has not been replaced with a wrong version. However, tampering with the page data 
 might go unnoticed. 
 
-The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
-be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
-length of `aad_metadata` is 512 bytes.
-
-Parquet-mr/-cpp implementations use the RBG-based IV construction as defined in the NIST 
-SP 800-38D document for the GCM ciphers (section 8.2.2).
+The initialization vectors (IVs) in AES GCM and CTR ciphers must be unique for each encryption key, 
+since a single re-use of an IV breaks the encryption. An IV in these two modes is often called a 
+nonce (“number used once”). Parquet encryption uses the RBG-based (random bit generator) IV 
+construction as defined in the NIST SP 800-38D document for the GCM ciphers (section 8.2.2). 
 
 
 ### AES_GCM_V1
-All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
-written after each ciphertext. The IVs (12 bytes) are written before each ciphertext.
+All modules are encrypted by an AES-GCM cipher, without padding. Unique IVs are generated and 
+written before each ciphertext. The IV length is 12 
+bytes (96 bits), which is generally considered to be the optimal size of GCM IVs.  The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” to be empty for 
+this IV length (see section 8.2.2). 
+
+The authentication tags (16 bytes) are written after each ciphertext.
 
 ### AES_GCM_CTR_V1
-Thrift modules are encrypted with the AES-GCM cipher, as described above. 
-The pages are encrypted with AES-CTR, where the IVs (16 bytes) are written before each 
-ciphertext.
+Thrift modules are encrypted with AES-GCM, as described above. The pages are encrypted by an AES-CTR
+cipher without padding. The nonces are generated and written before each ciphertext, with the nonce length of 16 bytes 
```

Comment:
Either "nonce" or "IV" should be used throughout the doc. Terminology should not change between GCM and CTR.